### ISSUE TITLE
The problem occurs try executing the file sample/sample_classification_with_augmentation.ipynb

### ISSUE THREAD

**Issue by lovejerryk52** [solislemuslab/tropical-stethoscope#6]
Dear authors of the animal sound classification project,

My name is Jerry Ho, I am currently a first-year student at university at Ho Chi Minh City, Vietnam.

I have read your paper "Classification of animal sounds in a hyperdiverse rainforest using Convolutional Neural Networks". I am very interested in it and wanted to try the performance of your proposed method. As guided, I tried to execute the file "sample/sample_classification_with_augmentation.ipynb" and encountered some issues that confused me:

Firstly, during training, it frequently early stopped at low epochs (around first 30 epochs).

Secondly, training up to 300 epochs is quite challenging, my team and I have not reached this number. The best is around 260 epochs and then early stopped.

Thirdly, after completing the training, I evaluated the model and got very bad results (best loss test loss, test acc: [6.361281394958496, 0.1666666716337204] precision: 0.027777777777777776,recall: 0.16666666666666666,f1 score: 0.047619047619047616)

You can take a look at my colab notebook to better understand the problem https://colab.research.google.com/drive/1kos24hq02w2xLrDVFCskTuppcxKPok90?usp=sharing
and if you want to discuss further, my email is coursetruong@gmail.com

During the experiment I used google colab and the Windows operating system.

I really need your help to resolve this issue.
Looking forward to hearing back from you.

I sincerely thank you for sharing the useful source code. I really appreciate it.
Please forgive me if my wording is not proper, as English is not my native language.

**Comment by crsl4** [comment]
Hello Jerry! I tag @YurenSUN who is the main developer of this pipeline.

**Comment by YurenSUN** [comment]
Hi Jerry,

Thank you for pointing out the errors within the pipeline! We are sorry that this unexpected issue happened!

It is expected to stop before 300 epochs and how many epochs it needs to train to reach a good model performance is based on the training samples selected. The reason is that as we use early stopping, the model will stop training when the validation loss no longer improves, which indicates that the model is not possibly generalizing so it might not be useful to continue training.

I took a quick look into the Colab that you provided but unsure about what is the cause for this. My intuition is that because we randomly picked up samples, it might result in us having bad luck and including bad samples in the validation and test sets, but there should possibly be other mistakes that I made when putting up this sample before. It could also be related to the sample files that I put together or the updates in TensorFlow versions that unexpectedly impact the pipeline. It is also encouraged to try out your own samples if you have any.

It’s a bit hard to quickly pick up on what went wrong as it was developed a long time before but I remembered that the pipeline was working two years ago. I might have more time to look deeper into it and get back to you in a week or two but not sure whether this might be too late or whether I might be occupied by other things later as my schedule is quite occupied by my university workloads.

Thank you so much for pointing out these errors and sorry that we are not able to help you resolve this quickly. Please let me know if there are any other things that we could help with.

Best,
Yuren

**Comment by lovejerryk52** [comment]
Dear YurenSUN,

Thank you very much for your prompt response. I really appreciate you getting back to me so quickly.
I sincerely need your help to resolve this issue. I hope to receive your assistance soon.

Best regards,
Jerry

**Comment by YurenSUN** [comment]
Hi Jerry,

Sorry to get back to you late. The broken pipeline might be because we used TensorFlow 1.x previously and there might be issues with using the generator function with TensorFlow 2.x (the latest versions).

I updated the sample Jupyter Notebook ([here](https://github.com/solislemuslab/tropical-stethoscope/blob/master/example/sample_classification_with_augmentation.ipynb)) with the Python scripts that we ran for experiments. [Here](https://drive.google.com/file/d/11z2p1cR7Lv9nCtlR7XKcdgbw9-BxwuRX/view?usp=sharing) is a copy of it with training and evaluation process, data, and metrics. 

As mentioned before, the process involves randomness so the results will possibly differ each time, and also feel free to change or play around with different configs. Besides, it is not necessarily that the training process will take 300 epochs each time as we could stop earlier when the validation loss no longer decreases, indicating that the model is not generalizing any longer. Also, feel free to play with different numbers for the "patient" parameter when setting up the early stopping or not using the early stop at all.

Thank you!

Best,
Yuren
